server/jetty/src/main/java/io/deephaven/server/jetty/jsplugin/JsPlugins.java

```diff
@@ -0,0 +1,112 @@
+
+    /**
+     * Creates a new js plugins instance with a temporary zip filesystem.
+     *
+     * @return the js plugins
+     * @throws IOException if an I/O exception occurs
+     */
+    public static JsPlugins create() throws IOException {
+        final Path tempDir = Files.createTempDirectory(JsPlugins.class.getName());
+        tempDir.toFile().deleteOnExit();
+        final Path fsZip = tempDir.resolve("fs.zip");
+        fsZip.toFile().deleteOnExit();
+        // Note, the URI needs explicitly be parseable as a directory URL ending in "!/", a requirement of the jetty
+        // resource creation implementation, see
+        // org.eclipse.jetty.util.resource.Resource.newResource(java.lang.String, boolean)
+        final URI uri = URI.create(String.format("jar:file:%s!/", fsZip));
+        final JsPlugins jsPlugins = new JsPlugins(uri);
+        jsPlugins.init();
+        return jsPlugins;
+    }
+
+    private final URI filesystem;
+    private final List<Map<String, String>> plugins;
+
+    private JsPlugins(URI filesystem) {
+        this.filesystem = Objects.requireNonNull(filesystem);
+        this.plugins = new ArrayList<>();
+    }
+
+    /**
+     * Creates a {@link DefaultServlet} servlet holder.
+     *
+     * @param name the name
+     * @return the servlet holder
+     */
+    public ServletHolder servletHolder(String name) {
+        final ServletHolder jsPlugins = new ServletHolder(name, DefaultServlet.class);
+        jsPlugins.setInitParameter("resourceBase", filesystem.toString());
+        jsPlugins.setInitParameter("pathInfoOnly", "true");
+        jsPlugins.setInitParameter("dirAllowed", "false");
+        // etag support picked up via io.deephaven.server.jetty.JettyBackedGrpcServer
+        // jsPlugins.setInitParameter("etags", "true");
+        jsPlugins.setAsyncSupported(true);
+        return jsPlugins;
+    }
+
+    @Override
+    public synchronized void register(JsPlugin jsPlugin) throws IOException {
+        for (Map<String, String> plugin : plugins) {
+            if (jsPlugin.name().equals(plugin.get(NAME))) {
+                throw new IllegalArgumentException(
+                        String.format("js plugin with name '%s' already exists", jsPlugin.name()));
+            }
+        }
+        try (final FileSystem fs = FileSystems.newFileSystem(filesystem, Map.of())) {
+            final Path pluginPath = fs.getPath("/", jsPlugin.name());
+            jsPlugin.copyTo(pluginPath);
+            plugins.add(Map.of(NAME, jsPlugin.name(), VERSION, jsPlugin.version(), MAIN, jsPlugin.main()));
+            writeManifest(fs);
+        }
+    }
+
+    private void init() throws IOException {
+        try (final FileSystem fs = FileSystems.newFileSystem(filesystem, Map.of("create", "true"))) {
+            writeManifest(fs);
+        }
+    }
+
+    private void writeManifest(FileSystem fs) throws IOException {
+        final Path manifestPath = fs.getPath("/", MANIFEST_JSON);
```

Comment:
using "/" violates the assumption you had expressed in chat, where this was going to be testable on a plain fs, rather than only jimfs or zip?

Reply:
Correct - given that zipfs works on new filesystems and has semantics required closing the filesystem, it didn't mesh cleanly w/ how one might use jimfs or plain fs. (Note: this does *not* prevent `JsPlugin` from being local filesystem, or something else; just effects the registration implementation.)

As such, I should probably rename and document this *class* as working with zip filesystems explicitly. If we find we need other implementations centered aroundjimfs or localfs in the future, we can create a new `JsPluginRegistration`.